storage/inkless/src/main/java/io/aiven/inkless/cache/CaffeineCache.java

```diff
@@ -37,34 +43,49 @@ public void close() throws IOException {
 
     @Override
     public FileExtent computeIfAbsent(final CacheKey key, final Function<CacheKey, FileExtent> mappingFunction) {
-        return cache.asMap().computeIfAbsent(key, mappingFunction);
+        final CompletableFuture<FileExtent> future = new CompletableFuture<>();
+        final CompletableFuture<FileExtent> existingFuture = cache.asMap().computeIfAbsent(key, (cacheKey) -> {
+            return future;
+        });
+        // Use Object.equals, if existing future is not the same object as created in this function
+        // there was a pending cache load and this call is required to join the existing.
+        if (!future.equals(existingFuture)) {
+            return existingFuture.join();
+        }
+        final FileExtent fileExtent = mappingFunction.apply(key);
+        future.complete(fileExtent);
+        return fileExtent;
```

Comment:
What prevents us from doing this?
```suggestion
        final CompletableFuture<FileExtent> existingFuture = cache.asMap().computeIfAbsent(key, (cacheKey) -> {
            return CompletableFuture.supplyAsync(() -> mappingFunction.apply(key));
        });
        return existingFuture.join();
```

AFAIU, there's `ConcurrentHashMap` inside, so it's guaranteed that the function will be called once if the key doesn't exist.